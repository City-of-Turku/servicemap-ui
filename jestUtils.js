import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { ThemeProvider } from '@mui/material/styles';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import themes from './src/themes';
import finnish from './src/i18n/fi';

const mockStore = configureStore([]);

// Mock props for intl provider
const intlMock = {
  locale: 'fi',
  messages: finnish,
  wrapRichTextChunksInFragment: false,
};

const Providers = (mockState) => ({ children }) => {
  const store = mockStore(mockState);
  return (
    <Provider store={store}>
      <IntlProvider {...intlMock}>
        <ThemeProvider theme={themes.SMTheme}>
          {children}
        </ThemeProvider>
      </IntlProvider>
    </Provider>
  );
}

export const getRenderWithProviders = mockState => component => render(component, { wrapper: Providers(mockState) });