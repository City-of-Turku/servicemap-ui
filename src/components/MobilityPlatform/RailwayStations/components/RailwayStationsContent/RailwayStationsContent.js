import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Typography } from '@mui/material';
import styled from '@emotion/styled';
import { format } from 'date-fns';
import { useIntl } from 'react-intl';
import { fetchRailwaysData } from '../../../mobilityPlatformRequests/mobilityPlatformRequests';
import {
  StyledContainer, StyledHeaderContainer, StyledFlexContainer, StyledTextContainer,
} from '../../../styled/styled';

const RailwayStationsContent = ({ item, stationsData }) => {
  const [stationTrainsData, setStationTrainsData] = useState([]);

  const intl = useIntl();
  const formatDateTime = timeValue => format(new Date(timeValue), 'HH:mm');

  const optionsToParams = options => {
    const params = new URLSearchParams();
    Object.entries({ ...options }).forEach(([key, value]) => {
      params.set(key, value);
    });
    return params.toString();
  };

  useEffect(() => {
    const options = {
      minutes_before_departure: 180,
      minutes_after_departure: 15,
      minutes_before_arrival: 180,
      minutes_after_arrival: 15,
      train_categories: 'Long-distance',
    };
    const endpoint = `live-trains/station/${item.stationShortCode}`;
    const params = optionsToParams(options);
    const query = `${endpoint}?${params}`;
    fetchRailwaysData(query, setStationTrainsData);
  }, [item.stationShortCode]);

  const filterArrivals = data => data.filter(train => {
    const lastRow = train.timeTableRows.slice(-1)[0];
    return lastRow.stationShortCode === item.stationShortCode && lastRow.type === 'ARRIVAL';
  });

  const filterDeparting = data => data.filter(train => {
    const firstRow = train.timeTableRows[0];
    return firstRow.stationShortCode === item.stationShortCode && firstRow.type === 'DEPARTURE';
  });

  const arrivingTrains = filterArrivals(stationTrainsData);
  const departingTrains = filterDeparting(stationTrainsData);

  const findStation = (stationsArr, codeValue) => stationsArr.find(station => station.stationShortCode === codeValue);

  const renderDestination = data => {
    const lastIdx = data.slice(-1)[0];
    const arrivalStation = findStation(stationsData, lastIdx.stationShortCode);
    return (
      <StyledText variant="body2" component="p">
        {arrivalStation.stationName}
      </StyledText>
    );
  };

  const renderTrainInfo = train => (
    <StyledText variant="body2" component="p">{`${train.trainType} ${train.trainNumber}`}</StyledText>
  );

  const renderTimeValues = elem => (
    <StyledText key={elem.scheduledTime} variant="body2" component="p">
      {elem.liveEstimateTime && elem.differenceInMinutes > 1
        ? `${formatDateTime(elem.liveEstimateTime)} (${formatDateTime(elem.scheduledTime)})`
        : `${formatDateTime(elem.scheduledTime)}`}
    </StyledText>
  );

  return (
    <StyledContainer>
      <StyledHeaderContainer>
        <StyledText variant="subtitle1" component="h4">
          {item?.stationName}
        </StyledText>
      </StyledHeaderContainer>
      <div>
        <div>
          <StyledTextContainer>
            <StyledText variant="subtitle1" component="h5">
              {intl.formatMessage({ id: 'mobilityPlatform.content.departingTrains.title' })}
            </StyledText>
            {!departingTrains?.length ? (
              <StyledText variant="body2" component="p">
                {intl.formatMessage({ id: 'mobilityPlatform.content.departingTrains.empty' })}
              </StyledText>
            ) : null}
          </StyledTextContainer>
          {departingTrains?.map(train => (
            <StyledFlexContainer key={train.trainNumber}>
              <TrainIcon color="rgba(7, 44, 115, 255)" className="icon-icon-hsl-train" />
              {renderTrainInfo(train)}
              {renderDestination(train.timeTableRows)}
              {train.timeTableRows
                .filter(elem => elem.stationShortCode === item.stationShortCode && elem.type === 'DEPARTURE')
                .map(elem => renderTimeValues(elem))}
            </StyledFlexContainer>
          ))}
        </div>
        <div>
          <StyledTextContainer>
            <StyledText variant="subtitle1" component="h5">
              {intl.formatMessage({ id: 'mobilityPlatform.content.arrivingTrains.title' })}
            </StyledText>
            {!arrivingTrains?.length ? (
              <StyledText variant="body2" component="p">
                {intl.formatMessage({ id: 'mobilityPlatform.content.arrivingTrains.empty' })}
              </StyledText>
            ) : null}
          </StyledTextContainer>
          {arrivingTrains?.map(train => (
            <StyledFlexContainer key={train.trainNumber}>
              <TrainIcon color="rgba(7, 44, 115, 255)" className="icon-icon-hsl-train" />
              {renderTrainInfo(train)}
              {renderDestination(train.timeTableRows)}
              {train.timeTableRows
                .filter(elem => elem.stationShortCode === item.stationShortCode && elem.type === 'ARRIVAL')
                .map(elem => renderTimeValues(elem))}
            </StyledFlexContainer>
          ))}
        </div>
      </div>
    </StyledContainer>
  );
};

const TrainIcon = styled.span(({ color }) => ({
  fontSize: 20,
  width: 20,
  height: 20,
  lineHeight: '21px',
  marginLeft: 6,
  marginRight: 4,
  color,
}));

const StyledText = styled(Typography)(({ theme }) => ({
  marginBottom: theme.spacing(0.5),
}));

RailwayStationsContent.propTypes = {
  item: PropTypes.shape({
    stationShortCode: PropTypes.string,
    stationName: PropTypes.string,
  }),
  stationsData: PropTypes.arrayOf(
    PropTypes.shape({
      stationShortCode: PropTypes.string,
    }),
  ),
};

RailwayStationsContent.defaultProps = {
  item: {},
  stationsData: [],
};

export default RailwayStationsContent;
