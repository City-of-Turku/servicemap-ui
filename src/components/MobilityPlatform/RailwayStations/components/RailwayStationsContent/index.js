import { injectIntl } from 'react-intl';
import RailwayStationsContent from './RailwayStationsContent';

export default injectIntl(RailwayStationsContent);
