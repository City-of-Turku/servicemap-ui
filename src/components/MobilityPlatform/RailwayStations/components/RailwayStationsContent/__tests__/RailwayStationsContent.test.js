// Link.react.test.js
import React from 'react';
import { getRenderWithProviders } from '../../../../../../../jestUtils';
import finnishTranslations from '../../../../../../i18n/fi';
import RailwayStationsContent from '../index';

const mockProps = {
  item: {
    stationName: 'Kupittaa',
    stationShortCode: 'KUT',
  },
  stationsData: [
    {
      stationName: 'Kupittaa',
      stationShortCode: 'KUT',
    },
    {
      stationName: 'Turku asema',
      stationShortCode: 'TKU',
    },
    {
      stationName: 'Turku Satama',
      stationShortCode: 'TUS',
    },
    {
      stationName: 'Helsinki asema',
      stationShortCode: 'HKI',
    },
    {
      stationName: 'Tampere asema',
      stationShortCode: 'TRE',
    },
  ],
};

const renderWithProviders = getRenderWithProviders({});

describe('<RailwayStationsContent />', () => {
  it('should match snapshot', () => {
    const { container } = renderWithProviders(<RailwayStationsContent {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('does show text', () => {
    const { container } = renderWithProviders(<RailwayStationsContent {...mockProps} />);
    const p = container.querySelectorAll('p');
    const h4 = container.querySelector('h4');
    const h5 = container.querySelectorAll('h5');
    expect(h4.textContent).toContain(mockProps.item.stationName);
    expect(h5[0].textContent).toContain(finnishTranslations['mobilityPlatform.content.departingTrains.title']);
    expect(h5[1].textContent).toContain(finnishTranslations['mobilityPlatform.content.arrivingTrains.title']);
    expect(p[0]).toBeInTheDocument();
    expect(p[1]).toBeInTheDocument();
  });
});
