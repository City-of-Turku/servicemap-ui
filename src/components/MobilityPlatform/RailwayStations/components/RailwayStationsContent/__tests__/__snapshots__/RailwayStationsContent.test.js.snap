// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RailwayStationsContent /> should match snapshot 1`] = `
<div>
  <div
    class="css-1ynyhby"
  >
    <div
      class="css-1dznbe2"
    >
      <h4
        class="MuiTypography-root MuiTypography-subtitle1 css-1im1a93-MuiTypography-root"
      >
        Kupittaa
      </h4>
    </div>
    <div>
      <div>
        <div
          class="css-1fhgjcy"
        >
          <h5
            class="MuiTypography-root MuiTypography-subtitle1 css-1im1a93-MuiTypography-root"
          >
            L채htev채t junat
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body2 css-qabfq1-MuiTypography-root"
          >
            Ei l채htevi채 junia
          </p>
        </div>
      </div>
      <div>
        <div
          class="css-1fhgjcy"
        >
          <h5
            class="MuiTypography-root MuiTypography-subtitle1 css-1im1a93-MuiTypography-root"
          >
            Saapuvat junat
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body2 css-qabfq1-MuiTypography-root"
          >
            Ei saapuvia junia
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
